src/CodeGen_LLVM.cpp

```diff
@@ -1830,17 +1872,21 @@ Value *CodeGen_LLVM::codegen_buffer_pointer(const string &buffer, Halide::Type t
 }
 
 Value *CodeGen_LLVM::codegen_buffer_pointer(Value *base_address, Halide::Type type, Value *index) {
-    llvm::Type *base_address_type = base_address->getType();
-    unsigned address_space = base_address_type->getPointerAddressSpace();
-
     type = upgrade_type_for_storage(type);
+    llvm::Type *load_type = llvm_type_of(type);
+    unsigned address_space = base_address->getType()->getPointerAddressSpace();
+    llvm::Type *pointer_load_type = load_type->getPointerTo(address_space);
 
-    llvm::Type *load_type = llvm_type_of(type)->getPointerTo(address_space);
+#if LLVM_VERSION <= 150
+    llvm::Type *base_address_type = base_address->getType();
 
     // If the type doesn't match the expected type, we need to pointer cast
-    if (load_type != base_address_type) {
-        base_address = builder->CreatePointerCast(base_address, load_type);
+    if (pointer_load_type != base_address_type) {
```

Comment:
Does this apply to the branched-but-not-final LLVM14 as well? (As of right now we're not testing that but I will set up the buildbots to add that shortly)